### ISSUE TITLE
A case of questionable unit prefix choice in OMEdit plots

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#12759]
### Description
This report is different from #12650.
I'm not sure if this is a duplicate of other tickets. In case it is, please excuse me and delete it.

Consider this screenshot: 
![image](https://github.com/user-attachments/assets/858bc19d-397a-4a41-866f-8908316e3794)
When the two curves are together, the choice of kNm for the blue curve makes it hardly distinguishable. In this case it would be better to use for both of Nm.

### Steps to Reproduce
I obtained this from a rather long model. In case of need I can supply it to allow reproduction.

### Version and OS
OM 1.23.1

**Comment by casella** [comment]
@ceraolo what to do with unit prefixing when you have multiple curves in the same plot is a bit controversial.

One example I recently ran into was analyzing the performance of a drone model. I was reading the angles around the three axes along some maneuvers which were supposed to only affect one angle, so when plotting them I would have liked to see one curve moving around by several degrees and the other two ones remaining virtually at zero. Instead, I also got the other two moving around in the plot, but then their units were like udeg or ndeg.

So, there are cases when the multiple should be preferably the same for all the curves, and others where they should be different. I'm not sure what is the best way to handle this in the GUI.

**Comment by ceraolo** [comment]
> So, there are cases when the multiple should be preferably the same for all the curves, and others where they should be different. I'm not sure what is the best way to handle this in the GUI.

However, this ticket's case should not be controversial. The purpose of plots is to see the curve behaviours. Using different prefixes for different curves in this case just reduces legibility.
I mean, switching from uniform prefixes to split prefixes (as in this case, one unit with prefix, another without) should be justified by a neat increase in legibility.

**Comment by casella** [comment]
> For me the best solution is to always use the same prefix for all curves, if they are chosen automatically.

I agree with that, and we can start implementing this right away.

> More complex cases should be dealt with in a different way, e.g.:
> 
>     1. adding the possibility to plot matematical combinations of plots (it is possible since many years ago in my PlotXY)
> 
>     2. adding the possibility for the user to manually choose independent prefixes for individual curves.
> 
> 
> In case there were some work-force invest in these things, was I in charge to decide, I would spend that working force with priority in the direction of point 1 above, instead of 2.

Regarding 2., I think this should already possible, since you have a displayUnit drop-down menu in the Variables browser. Unfortunately, it seems that it doesn't support prefixes, I simulated Modelica.Electrical.Analog.Examples.CauerLowPassAnalog and tried to change the display units of some voltages, but I only got "V" in the drop-down list.

@adeas31 could we provide all prefixes there as we already do for parameter inputs?

Regarding 1. it is an interesting proposal, but we have a lot of higher-priority work to do first. Why not trying to integrate your tool with OMEdit instead?

**Comment by ceraolo** [comment]
> Regarding 1. it is an interesting proposal, but we have a lot of higher-priority work to do first. Why not trying to integrate your tool with OMEdit instead?

For me it would be nice.
It a class is based on Qt 5, making usage of regExpr (not regularExpressions).
I could polish it, and its documentation and prepare a small demo video

**Comment by casella** [comment]
Excellent, please coordinate with @adeas31

**Comment by ceraolo** [comment]
> Excellent, please coordinate with @adeas31

I will come back to adeel on late August, to check with him if there is feasibility.

**Comment by adeas31** [comment]
> For me the best solution is to always use the same prefix for all curves, if they are chosen automatically.

Please see #11201. We use to have same prefix for all curves and it was chosen automatically. As part of #11201 we decided to change that functionality and implemented automatic individual prefixes.

> @adeas31 could we provide all prefixes there as we already do for parameter inputs?

See #12222.

Regarding PlotXY integration, we talked about it in the past. If we want to replace it with OMPlot then it should at least have all the functionalities of OMPlot.

**Comment by casella** [comment]
> > For me the best solution is to always use the same prefix for all curves, if they are chosen automatically.
> 
> Please see #11201. We use to have same prefix for all curves and it was chosen automatically. As part of #11201 we decided to change that functionality and implemented automatic individual prefixes.

In the case of #11201 we actually had variables with different base units, one was W and the other one was m/s. In that case, of course, using the same prefix was not a good idea, as the two variables were completely unrelated, and even had different units.

When you have two (or more) variables with the same base unit, you often need to compare them. The question is: how?

Consider @ceraolo's original test case:  tau is 1000 N/m, which is rendered (correctly) as 1 kN/m, while tauGenerated is slightly lower, so it is rendered (correctly) as hundreds of N/m. If you display them separately, this is exactly what we want. However, if you want to compare them, using different prefixes is of course not a good idea, because tau (which is larger) looks much smaller than tauGenerated, and in fact it is impossible to read its actual value, since a value of 1 is completely lost on a scale going from 0 to 1000.

The other case that I mentioned in my previous comment is reproduced by this MWE:
```modelica
model M
  Real x[3](each unit="m");
equation
  x[1] = 3*sin(30*time);
  x[2] = 3*cos(30*time);
  x[3] = 20e-9*cos(30*time-0.02);
end M;
```
In this case you have two components with output of the order of a meter, while the third one is several microns, e.g. because of numerical errors. When you display this, you get:
![immagine](https://github.com/user-attachments/assets/cb9b9c4f-955d-4a21-8db0-e41c2e8eddcb)
which is quite confusing, as at first sight it seems that the third component is larger than the first two. From this point of view, if you want to figure out what is the actual 3D behaviour, you should get the same unit (including the prefix) for all three components:
![immagine](https://github.com/user-attachments/assets/4dc3d3ba-2091-45b1-8aca-f9554356b631)

In this case, I could obtain the last plot by disabling prefixes altogether, but this wouldn't work well if `x[1]` and `x[2]` were, e.g., in the range of km or something.

Bottom line: I'm not really sure what is the best GUI design regarding this issue. Maybe the most straightforward way is to have each variable handled separately by default, but on the other hand allow to change the unit prefix separately when right-clicking on each variable in the legend. Currently, the only option is to toggle unit prefixing on and off, but it is not possible to manually set the prefix.

@ceraolo what do you think?

**Comment by casella** [comment]
> Regarding PlotXY integration, we talked about it in the past. If we want to replace it with OMPlot then it should at least have all the functionalities of OMPlot.

I think we should have both. Let users decide on a case-by-case basis which one suits them best.

**Comment by adeas31** [comment]
> Bottom line: I'm not really sure what is the best GUI design regarding this issue. Maybe the most straightforward way is to have each variable handled separately by default, but on the other hand allow to change the unit prefix separately when right-clicking on each variable in the legend. Currently, the only option is to toggle unit prefixing on and off, but it is not possible to manually set the prefix.

This is fine for me. Our current implementation handles each variable separately. We only need to fix #12222.

**Comment by max-privato** [comment]
> 
> Regarding PlotXY integration, we talked about it in the past. If we want to replace it with OMPlot then it should at least have all the functionalities of OMPlot.

I remember this well. 

I thought the idea of @casella here was diifferent: just try to integrate the feature to algebrically mix variables: sums, products (with possible use of parentheses), even use of trigonometric functions, and other things. Not total substitution of OMPlot.

Since this requires a non negligible effort from  both @adeas31 and me, we can do this only if there is plain support of the idea also by you, @adeas31.
In case there is a least a provisonal acceptance of the idea, we could have a short on-line meeting to try to envisage if this is sufficiently easy, fast and acceptable by you.
Otherwise, we can easily drop the idea.

**Comment by bilderbuchi** [comment]
> However, if you want to compare them, using different prefixes is of course not a good idea, because tau (which is larger) looks much smaller than tauGenerated, and in fact it is impossible to read its actual value, since a value of 1 is completely lost on a scale going from 0 to 1000.

Hence the proposal of #11218, which _would_ make it possible to read both traces.

IMHO, a sane default behaviour would be to scale traces such that their variation covers an _appreciable amount of y range_, so that their behaviour can be _read properly_. That means I'd expect the first variant in Francesco's sine example. "Looks"/relationships are secondary to useful display of value information here, IMO.

Assessments that one µm-scale trace should be displayed much smaller (uselessly small in this case), as it relates to another m-scale trace somehow, are not possible to make automatically I think. Therefore, this should be enabled by having manual adjustment possibilities beyond the autoscaling/unit-prefixing.

**Comment by casella** [comment]
> Assessments that one µm-scale trace should be displayed much smaller (uselessly small in this case), as it relates to another m-scale trace somehow, are not possible to make automatically I think. Therefore, this should be enabled by having manual adjustment possibilities beyond the autoscaling/unit-prefixing.

That's the current agreement, it just needs #12222 to be fixed.

**Comment by pierre-haessig** [comment]
I also got hit with this bug with students. Surely having #12222 would be useful.

Now about what should be the *default/starting* choice, the case I faced was similar to @ceraolo's, so I would favor having only one common prefix for variables sharing the same unit, but  I guess it makes the logic for prefix selection more complex. Also, my view may be biased by this example, and perhaps there may be cases where people would be happy get an automatic superimposition of km and µm.

**Comment by pierre-haessig** [comment]
I again got hit by this bug (OpenModelica 1.24.5), but I think there's also a different issue going on here:
- one thing is I expect that the voltage which reach 3000 V to *not be converted* to kV (or else convert all voltages to kV). This is discussed here
- but more annoyingly, the variable browser claims to display V, not kV. I didn't find the correct issue for that problem

![Image](https://github.com/user-attachments/assets/4bff922f-8615-439e-bc00-d6f5d89347a1)

**Comment by bilderbuchi** [comment]
> * one thing is I expect that the voltage which reach 3000 V to _not be converted_ to kV (or else convert all voltages to kV). This is discussed here

I'd be curious about the reasoning for that expectation --- If you end up with a plot displaying small voltages O(10 V) like `Ra.v`, and very large voltages O (1000 V) like `cfY3ph.va`, with all units in kV (or V, for that matter), you won't see much, because the smaller ones collapse to wiggle around the zero line. 
IMO, OMEdit is doing you a service here by scaling the voltages so that you can assess the detail of their time evolutions in a meaningful manner, no?
I'm curious why you seem to disagree.

> * but more annoyingly, the variable browser claims to display V, not kV. I didn't find the correct issue for that problem

AFAICT, everything is correct, here? The variable browser displays ~ -300 V, so 0.3 kV. From looking at the plot image, the magenta trace, at the right end of the plot, is very small, far below the 1 kV tick, could well be 300 V but it's hard to tell from a few pixels.
What am I missing? what do you think is going wrong here?

**Comment by ceraolo** [comment]
@pierre-haessig , @bilderbuchi , thank you for this further contribution.
I think here we are in the domain of personal preferences. Do we want to see the relative size of curves or see their trend? In case two quantities have different orders of magnitude, using a prefix on one of them alters their relative size, but enhances their trend visibility. This is what OMEdit did in the @pierre-haessig example. 

I think the answer to this question depends on purposes. 

Since I do not have a strong opinion on this, I prefer to concentrate to a case which is OMEdit really wrong: this ticket's example. Here, adding prefix to tauGenerated, instead of adding visibility, jeopardises it a lot!
So my priority is to have fixed this behaviour which (I suppose we agree on this) plain wrong; for the rest, both choices are acceptable, given that the user having a different preference can change some scale manually.

**Comment by pierre-haessig** [comment]
Thanks @ceraolo and @bilderbuchi for your feedback.

I agree to let aside the issue of personal preferences.

Sorry that my screenshot added more confusion, but I believe it is the exact same problem as @ceraolo's one: the different prefix choices, which makes sense individually, create visual confusion when superimposed (I didn't mentioned it took me  some minutes to notice the "kV" unit in the legend: first I though I was facing with a low voltage issue...).

Coming back to my point 2:
> - but more annoyingly, the variable browser claims to display V, not kV. I didn't find the correct issue for that problem

The issue is that, from what I used, the **Variable browser is used to select the display unit in the Plot**. But here I'm only proposed with the choice "V". (There is no screenshot of @ceraolo's variable browser to see what unit is displayed for the different tau variables, but perhaps the situation is similar?)

So in the end we perhaps need to split the questions:
- do we agree that the unit displayed in the Variable browser should be the same as in the Plot? (presently it's not)
- do we agree that if some "smart display logic" adds some automatically prefix to the plot, then this choice should be user-reversible? (presently it's not)

and then lays aside the question of user preference: when plotting variables *1) which shares the same unit* on *2) one single plot*, is it preferable to:
- to have the same prefix, so as to correctly display the magnitude differences?
- to different prefixes (range adjusted), so as to better display the individual dynamics?

And indeed as per @bilderbuchi comment, depending on the point of view, the second option may be viewed as preferable in some circumstances. However, I'm claiming that I see in this thread three user cases where this default choice create confusion (@ceraolo's tau plot, @casella m vs µm and mine).

**Comment by pierre-haessig** [comment]
> > * do we agree that the unit displayed in the Variable browser should be the same as in the Plot? (presently it's not)
> 
> Yes, absolutely. Also because, as I understand, the variable browser is the only place where I can pick a unit, I can't do that from the plot set-up window.
> 
> I understand this should already be the case in version 1.25.0. [@pierre-haessig](https://github.com/pierre-haessig) can you please provide the source code of a test case where this doesn't work for you?
> 

OK, I get the problem: I'm on 1.24.5, so that's why I cannot choose the unit.

> > * do we agree that if some "smart display logic" adds some automatically prefix to the plot, then this choice should be user-reversible? (presently it's not)
> 
> Yes, it is. If you don't like automatic unit prefixes, go to Tools | Options | Plotting | General | Prefix units and uncheck it

Thanks I unchecked to box and this solves the problem!

I agree with your suggestion to add another option "Use same prefix for all variables with the same unit". Only, there will need to be clarification as to how it interacts with the existing one "Prefix units".

By the way, the English label "Prefix units" isn't crystal clear (but the tooltip is great), but the French translation "Unités de Préfixe" is weirder. Is it true that in the English label, "Prefix" is meant to be an action verb?

**Comment by casella** [comment]
> I agree with your suggestion to add another option "Use same prefix for all variables with the same unit". Only, there will need to be clarification as to how it interacts with the existing one "Prefix units".

It will have to be a sub-choice: if you check Prefix units, you can also check or uncheck that option. If you uncheck Prefix units, it will be greyed out and have no effect.

**Comment by casella** [comment]
> By the way, the English label "Prefix units" isn't crystal clear (but the tooltip is great)

Labels need to be short by definition. The long explanation is meant to be provided by the tooltip hint. If you have a suggestion for an equally short but clearer label, we can change that easily. 

> but the French translation "Unités de Préfixe" is weirder. Is it true that in the English label, "Prefix" is meant to be an action verb?

I need to check what the status is with that.

**Comment by casella** [comment]
@adeas31 could you take care of adding the new option as explained [above](https://github.com/OpenModelica/OpenModelica/issues/12759#issuecomment-2741292130)? Thanks!

**Comment by ceraolo** [comment]
> * to add one more option under Tools | Options | Plotting | General | Prefix units, e.g. "Use same prefix for all variables with the same unit" - this gets everyone happy

Good!
The case in this ticket description would probably still fool the internal algorithm when this option is unchecked.
But maybe is a too special case, for which it is unjustified to try to find a special rule to manage.

**Comment by pierre-haessig** [comment]
> > * do we agree that the unit displayed in the Variable browser should be the same as in the Plot? (presently it's not)
> 
> Yes, absolutely. Also because, as I understand, the variable browser is the only place where I can pick a unit, I can't do that from the plot set-up window.
> 
> I understand this should already be the case in version 1.25.0. [@pierre-haessig](https://github.com/pierre-haessig) can you please provide the source code of a test case where this doesn't work for you?

This morning I just got the update to 1.25.0~dev.beta.1-1-g5e8bf4a (Ubuntu 22.04) and made a new test with the following MWE:

```Modelica
model VoltageScaling "Ramping voltages with two different order of magnitudes"
  Real v1(unit="V");
  Real v2(unit="V");
equation
  v1 = time;
  v2 = 1e-3*time;
end VoltageScaling;
```

With the "Prefix Units" options turned back on I get:
- automatic prefix works (V → mV) in that case (albeit of course with the issue that mV gets superimposed with V ;-) )
- but the mV choice is neither _displayed_ nor _available_ in the diplayUnit dropdown menu of the variable browers.

**Comment by pierre-haessig** [comment]
> > By the way, the English label "Prefix units" isn't crystal clear (but the tooltip is great)
> 
> Labels need to be short by definition. The long explanation is meant to be provided by the tooltip hint. If you have a suggestion for an equally short but clearer label, we can change that easily.

OK this gets quite out of topic, sorry, but to keep things at the same place, here are my suggestions:
- "Préfixer automatiquement les unités", or shorter if needed "Préfixer les unités". The verb "préfixer" is quite rare but I checked that it exists in French (to be sure it's not just engineers' dialect ;-) )
- or "Unités avec préfixe" which is closer to the present translation

**Comment by casella** [comment]
@adeas31, maybe I now understand what the problem is: when automatic unit prefixing is performed on a plot, we should update the display unit accordingly.

Also, when setting parameters with basic units in the parameter editing windows, we get all the prefixed units automatically. I think we should do the same also in the variable browser.

Can we fix this in time for 1.25.0? Thanks!

**Comment by bilderbuchi** [comment]
> we should update the display unit accordingly.

not in the source code, though, right?

**Comment by pierre-haessig** [comment]
> > we should update the display unit accordingly.
> 
> not the displayunit in the source code, though, right?

Indeed, the displayUnit in the Variable browser. Here is the screenshot of my [VotlageScaling example above](https://github.com/OpenModelica/OpenModelica/issues/12759#issuecomment-2742768065) (made with OM `1.24.5`, but reproduced with `1.25.0~dev.beta.1-1-g5e8bf4a`)

![Image](https://github.com/user-attachments/assets/5590ac42-e253-466d-95b5-0064cab177cf)

**Comment by casella** [comment]
See also #12222. Probably this ticket is now just a duplicate of that.

**Comment by pierre-haessig** [comment]
> See also [#12222](https://github.com/OpenModelica/OpenModelica/issues/12222). Probably this ticket is now just a duplicate of that.

Thanks for the pointer.

I would say that #12222 is a larger feature request in the sense that making available *all* prefixes in the Variable browser will in particular make available the *one prefix used in the plot*. So in the end solving #12222 will solve the problem with `VoltageScaling` example https://github.com/OpenModelica/OpenModelica/issues/12759#issuecomment-2747295018.

**Comment by casella** [comment]
Yes. I would go for fixing #12222 outright.

**Comment by adeas31** [comment]
> > > we should update the display unit accordingly.
> > 
> > 
> > not the displayunit in the source code, though, right?
> 
> Indeed, the displayUnit in the Variable browser. Here is the screenshot of my [VotlageScaling example above](https://github.com/OpenModelica/OpenModelica/issues/12759#issuecomment-2742768065) (made with OM `1.24.5`, but reproduced with `1.25.0~dev.beta.1-1-g5e8bf4a`)
> 
> ![Image](https://github.com/user-attachments/assets/5590ac42-e253-466d-95b5-0064cab177cf)

When you plot `v2` then we automatically add the prefix unit based on the values of `v2` which in this case is `mV`. As I understand you want to see `mV` in Variables Browser as well when you plot `v2`.

This will cause a cyclic behaviour. Even fixing #12222 will also cause a cycle. When we set `mV` in Variables Browser and then user again changes the display unit in the Variables Browser to `V` then we end up in a cycle since the automatic calculation changes it to `mV`. So we are doing something like this, Variables Browser->Plot->Variables Browser->Plot and so on.....

There are 2 solutions either **we don't add automatic prefix and let the user do it explicitly since we will list all the prefix units in the drop down list of Variables Browser** OR **do not change the display unit in the Variables Browser**. With the later solution what I mean is that when you select `v2` you will see `mV` in the plot and `V` in the Variables Browser. We only use Variables Browser display unit when adding/updating a plot curve and don't change it afterwards.

Note: Dymola does not automatically add the prefix unit instead it just lists all the prefix units and let the user manually select what he wants to plot.

**Comment by casella** [comment]
Thanks @adeas31 for making this clear. I think we could break the loop, though.

1. When first plotting the diagram, if there is no displayUnit set, the unit is given,  the prefix is automatically selected (if the prefix unit option is active) and the drop-down menu is automatically updated
2. If the user then changes the unit from the drop-down menu, that selection takes over the automatic selection, which is overridden

Would that work? I guess you have to change the UI chain of events, but the resulting behaviour just makes sense to me. You will have two kind of users:
- those who don't like auto unit prefixing will disable it, and then just pick the right display unit manually (as in Dymola)
- those who like auto unit prefixing will enable it, they will get some auto choice (the same for all plots or not depending on their choice) but will still be able to change it manually as they see fit

**Comment by pierre-haessig** [comment]
> Thanks [@adeas31](https://github.com/adeas31) for making this clear. I think we could break the loop, though.
> 
>     1. When first plotting the diagram, if there is no displayUnit set, the unit is given,  the prefix is automatically selected (if the prefix unit option is active) and the drop-down menu is automatically updated
> 
>     2. If the user then changes the unit from the drop-down menu, that selection takes over the automatic selection, which is overridden
> 
> 
> Would that work? I guess you have to change the UI chain of events, but the resulting behaviour just makes sense to me.

Agreed: once #12222 is implemented, the injection of the "automatic prefix choice" should mimic the user input, that is being performed in the Variable browser. No more hidden side channel! Thanks @adeas31  for pointing out this potential bug.

**Comment by casella** [comment]
See also #11834

**Comment by adeas31** [comment]
I have now finally pushed in #13789. Hopefully I have covered all the cases.

The variables browser now display the same units as in parameters dialog. When the user plots a variable and if prefix units is enabled then an automatic prefix will be added if needed and the variables browser is updated automatically. If the user explicitly change the `displayUnit` from the variables browser then the automatic prefix is not applied since the user manually changed the `displayUnit`.

@pierre-haessig I also updated the french translation from `Unités de Préfixe` to `Préfixer les unités`.

Please test tomorrow nightly build.
